library/core/src/iter/adapters/intersperse.rs

```diff
@@ -173,8 +173,7 @@ where
 {
     let mut accum = init;
 
-    // Use `peek()` first to avoid calling `next()` on an empty iterator.
-    if !needs_sep || iter.peek().is_some() {
+    if !needs_sep {
         if let Some(x) = iter.next() {
             accum = f(accum, x);
         }
```

Comment:
Looks like the idea behind `.peek().is_some()` here was to prevent calling fold() after next() returned None, as we don't know if `iter` is a `FusedIterator` or not. (It might start giving elements again after a `None`.)

However, it didn't work because `||` was used instead of `&&`, also causing the bug you're fixing with this PR.

Example that goes wrong both before and after this change:

```rust
#![feature(iter_intersperse)]

fn main() {
    struct A(i32); // Produces: None, Some(2), Some(3), None, ...
    impl Iterator for A {
        type Item = i32;
        fn next(&mut self) -> Option<i32> {
            self.0 += 1;
            if self.0 % 3 == 1 {
                None
            } else {
                Some(self.0)
            }
        }
    }

    dbg!(A(0).intersperse(1000).fold(0, |a, b| a + b));
}
```

This should give `0`, but gives `2005`, as it continued after the initial `None`.

This can be fixed by either:

1. Putting `Peekable` back and using `&& iter.peek().is_some()`; or
2. Adding a `} else { return accum; }` after the `if let Some(x) ..`:

```suggestion
    if !needs_sep {
        if let Some(x) = iter.next() {
            accum = f(accum, x);
        } else {
            return accum;
        }
```

Reply:
I blindly missed that. Thanks for the your thorough review. Will update the PR soon.